templating/secrets.go

```
@@ -0,0 +1,126 @@
+}
+
+// ResolveSecrets returns a list of resolved secrets and returns error if there is any failure in resolving a secret.
+func (secretResolver *SecretResolver) ResolveSecrets(ctx context.Context, secrets []*graph.Secret) (Values, error) {
+
+	resolvedSecrets := Values{}
+
+	if len(secrets) == 0 {
+		return resolvedSecrets, nil
+	}
+
+	// We will resolve in batches of 5 to avoid throttling errors on the vault providers
+	batchSize := 5
+	errorChan := make(chan error)
+	for index := 0; index < len(secrets); index += batchSize {
+		endIndex := index + batchSize
+
+		if endIndex > len(secrets) {
+			endIndex = len(secrets)
+		}
+
+		var resolvedChans []chan graph.SecretValue
+
+		for _, secret := range secrets[index:endIndex] {
+			// Validate secret before resolving
+			err := secret.Validate()
+			if err != nil {
+				if secret.ID == "" {
+					return resolvedSecrets, err
+				}
+				return resolvedSecrets, errors.Wrap(err, fmt.Sprintf("failed to validate secret with ID: %s", secret.ID))
+			}
+
+			if secret == nil {
+				continue
+			}
+
+			if secret.ResolvedChan == nil {
+				secret.ResolvedChan = make(chan graph.SecretValue)
+			}
+			resolvedChans = append(resolvedChans, secret.ResolvedChan)
+			go secretResolver.Resolve(ctx, secretResolver.azureEnvironment.KeyVaultEndpoint, secret, errorChan)
+		}
+
+		// Block until either:
+		// - The global context expires
+		// - Resolving a secret has error
+		// - All secrets are resolved successfully
+		for _, ch := range resolvedChans {
+			select {
+			case <-ctx.Done():
+				return resolvedSecrets, ctx.Err()
+			case secretValue := <-ch:
+				resolvedSecrets[secretValue.ID] = secretValue.Value
+			case err := <-errorChan:
+				return resolvedSecrets, err
+			}
+		}
+	}
+
+	return resolvedSecrets, nil
+}
+
+func resolveSecret(ctx context.Context, azureVaultResourceURL string, secret *graph.Secret, errorChan chan error) {
+	if secret == nil {
+		errorChan <- errors.New("secret cannot be nil")
```

what's the point of the channel, why not just return an error?